core/domain/story_services.py

```diff
@@ -200,31 +200,49 @@ def _save_story(committer_id, story, commit_message, change_list):
 
     story.validate()
     # Validate that all explorations referenced by the story exist.
-    exp_ids = []
-    for node in story.story_contents.nodes:
-        if node.exploration_id is not None:
-            exp_ids.append(node.exploration_id)
-    exp_summaries = (
-        exp_fetchers.get_exploration_summaries_matching_ids(exp_ids))
-
-    exp_summaries_dict = {
-        exp_id: exp_summaries[ind] for (ind, exp_id) in enumerate(exp_ids)
-    }
+    exp_ids = [
+        node.exploration_id for node in story.story_contents.nodes
+        if node.exploration_id is not None]
+
+    # The first exp ID in the story to compare categories later on.
+    sample_exp_id = exp_ids[0] if exp_ids else None
+
+    # Strict = False, since the existence of explorations is checked below.
+    exps_dict = (
+        exp_fetchers.get_multiple_explorations_by_id(exp_ids, strict=False))
+
     for node in story.story_contents.nodes:
         if (node.exploration_id is not None) and (
-                not exp_summaries_dict[node.exploration_id]):
+                node.exploration_id not in exps_dict):
             raise utils.ValidationError(
                 'Expected story to only reference valid explorations, '
                 'but found an exploration with ID: %s (was it deleted?)' %
                 node.exploration_id)
 
-    if exp_summaries:
-        common_exp_category = exp_summaries[0].category
-        for summary in exp_summaries:
-            if summary.category != common_exp_category:
+    if exps_dict:
+        common_exp_category = exps_dict[sample_exp_id].category
+        for exp_id in exps_dict:
+            exp = exps_dict[exp_id]
+            if exp.category != common_exp_category:
                 raise utils.ValidationError(
                     'All explorations in a story should be of the '
-                    'same category.')
+                    'same category. The explorations with ID %s and %s have'
+                    ' different categories.' % (sample_exp_id, exp_id))
+            if exp.language_code != 'en':
+                raise utils.ValidationError(
+                    'All explorations in a story should be in English.')
+            if exp.param_specs or exp.param_changes:
+                raise utils.ValidationError(
+                    'Expected no exploration to have parameter values in'
+                    ' it. Invalid exploration: %s' % exp.id)
+            for state_name in exp.states:
+                state = exp.states[state_name]
+                if not state.interaction.is_supported_on_android_app():
+                    raise utils.ValidationError(
+                        'Expected explorations to only have the following '
+                        'interactions: %s. Invalid exploration: %s' %
+                        (feconf.VALID_INTERACTION_IDS_FOR_ANDROID, exp.id))
```

Comment:
I don't think you need to quote the whole list in feconf here, which can become quite long. Instead, state the name of the invalid interaction, or the corresponding state.

When writing these error messages, think of the debugging workflow, and provide enough information to make it easy for the developer who is debugging. With the current error message, the developer would need to go through the entire exploration and would be needing to guess what the relevant state is.

Reply:
Done